Programmatically create permanent session

jmilkiewicz opened the pull request:
My draft implementation in which we can potentially support creating permanent session via REST API and YAML file. 
In case of conflict what we have in DB wins.

pdambrauskas commented on server/src/main/java/com/exacaster/lighter/application/sessions/SessionService.java:
```diff
@@ -39,23 +41,32 @@ public SessionService(ApplicationStorage applicationStorage,
+        return createSession(params, sessionId, ApplicationType.SESSION);
+    }
+
+    public Application createSession(SubmitParams params, String sessionId, ApplicationType applicationType) {
```
this should probably be private, otherwise we are able to create a new batch application through `createSession` method :)

pdambrauskas commented on server/src/main/java/com/exacaster/lighter/rest/PermanentSessionController.java:
```diff
@@ -0,0 +1,43 @@
+
+@Validated
+@Tags(@Tag(name = "Sessions"))
+@Controller("/lighter/api/permanentSessions")
```
I'm not sure if we should have a dedicated controller & ApplicationType for Permanent sessions.

Maybe it would make sense to have a single type "SESSION", as we do now, and just extend "SubmitParams" body for session creation with some optional "isPermanent" field? In that case, a single API controller would be enough.

To manage Permanent sessions, we could introduce a separate database table, something like "application_features", which would have reference to the application, as well some "management type = PERMANENT" field. That table could be reused for other similar features in the future, for example retry policies, repeat policies for the batch jobs, etc..

/cc @Minutis

jmilkiewicz replied:
> I'm not sure if we should have a dedicated controller & ApplicationType for Permanent sessions.
> 
> Maybe it would make sense to have a single type "SESSION", as we do now, and just extend "SubmitParams" body for session creation with some optional "isPermanent" field? In that case, a single API controller would be enough.
> 
Can also be done and then have an extra "if" to decide what to do. Please note that creating perm session is pretty simple and does not differ a lot from creating a regular session. Deleting it could require some extra logic (soft delete) and it would require extra "if". Why not make it in a controller ? 

> To manage Permanent sessions, we could introduce a separate database table, something like "application_features", which would have reference to the application, as well some "management type = PERMANENT" field. That table could be reused for other similar features in the future, for example retry policies, repeat policies for the batch jobs, etc..
YAGNI
> /cc @Minutis

pdambrauskas replied:
I do not think we need an extra “if” for deletion, the logics behind deletion should be the same for all applications. If we go with soft delete - we should add it to all types of applications.

if we go with a separate controllers for sessions & permanent sessions, we’d have to duplicate statement API under a different path, otherwise it would look quite strange, if you create a session through “/permannentSessions”, but post  statements to it through “/sessions/:id/statements”. Also some changes would have to be imlemented on the Lighter UI.

jmilkiewicz replied:
> I do not think we need an extra “if” for deletion, the logics behind deletion should be the same for all applications. If we go with soft delete - we should add it to all types of applications.

Ok. i assumed wrongly soft-delete would be only for permanent session - to find out if what we have in YAML has not been deleted before via REST API.

> if we go with a separate controllers for sessions & permanent sessions, we’d have to duplicate statement API under a different path, otherwise it would look quite strange, if you create a session through “/permannentSessions”, but post statements to it through “/sessions/:id/statements”. Also some changes would have to be imlemented on the Lighter UI.

true.

jmilkiewicz replied:
I added soft-delete and removed a dedicated controller for perm sessions

jmilkiewicz commented:
adding soft delete

jmilkiewicz commented on server/src/main/java/com/exacaster/lighter/application/sessions/SessionService.java:
```diff
@@ -95,10 +113,14 @@ public Optional<Application> fetchOne(String id) {
 
     public void deleteOne(String id) {
-        this.fetchOne(id).ifPresent(this::deleteOne);
+        fetchByType(id, ApplicationType.SESSION).ifPresent(this::deleteOne);
```
do we wanna  allow SessionService to remove all kind of application or just session ?

pdambrauskas replied:
It should remove only sessions.

jmilkiewicz replied:
done.

jmilkiewicz commented on server/src/main/java/com/exacaster/lighter/application/sessions/SessionHandler.java:
```diff
@@ -73,6 +78,72 @@ public void keepPermanentSessions() throws InterruptedException {
     }
 
+    @SchedulerLock(name = "keepPermanentSession", lockAtLeastFor = "1m")
+    @Scheduled(fixedRate = "1m", initialDelay = "2s")
```
I had to add initialDelay as it interfere with "handles soft deletes" test :(((((

pdambrauskas replied:
do we need both `keepPermanentSessions` & `keepPermanentSessions2`, looks like they are overlapping

jmilkiewicz replied:
nope. Will keep just one. Nevertheless micronaut from testing perspective is really bad if one can not easily disable schedulng or do proper test slices

jmilkiewicz replied:
only 1 keepPermanentSessions

jmilkiewicz commented:
@pdambrauskas I removed controller and added an extra param to SubmitParams. SubmitParams is used BOTH for 
- REST API param for session creation
- within `permanent-sessions` of YAML so the context is already defined ,ie it is permanent session

I am not really happy with this appraoch. I am not really familiar with micronaut but maybe to solve this there is sth like `https://projectlombok.org/features/experimental/Delegate` . It would allow us to do sth like:
```
class RESTSubmitParams(){
   @Delegate
   private final SubmitParams submitParams;

   private final Boolean permanent;

 
}
``` 

So from property binding all properties of SubmitParams would be "added" to RESTSubmitParams. Other option maybe would be to simply use inheritance:
```
class RESTSubmitParams extends SubmitParams(){
 private final Boolean permanent;
....
}
```

WDYT ?

pdambrauskas commented on server/src/main/java/com/exacaster/lighter/application/sessions/PermanentSession.java:
```diff
@@ -0,0 +1,26 @@
+
+public class PermanentSession {
+    private final Application application;
+    private final boolean isDeleted;
```
I think it is fine to add "isDeleted" directly to the `Application`, in that case `PermanentSession` class would not probably be needed. Or are there any more reasons for it to exist?

jmilkiewicz replied:
Ok, done

pdambrauskas commented on server/src/main/java/com/exacaster/lighter/storage/ApplicationStorage.java:
```diff
@@ -18,4 +20,8 @@ public interface ApplicationStorage {
     List<Application> findApplicationsByStates(ApplicationType type, List<ApplicationState> states, SortOrder order,
             Integer from, Integer size);
+
+    List<PermanentSession> findAllPermanentSessions();
```
I think it is nicer to not have `ApplicationType` specific methods here. Maybe we could add `findAllApplications(ApplicationType type)` instead?

jmilkiewicz replied:
done

pdambrauskas commented on server/src/main/java/com/exacaster/lighter/rest/SessionParameters.java:
```diff
@@ -0,0 +1,25 @@
+import java.util.Map;
+
+@Introspected
+public class SessionParameters extends SubmitParams {
```
maybe lets name it `SessionSubmitParams`, to align with `SubmitParams` name?

jmilkiewicz replied:
can do. nevertheless `SubmitParams` suggests parameters to be submitted to Spark session but in this case `SessionParameters ` is about submit parameters and lighter specific configuration of session. WDYT ?

pdambrauskas commented on server/src/main/java/com/exacaster/lighter/rest/SessionParameters.java:
```diff
@@ -0,0 +1,25 @@
+public class SessionParameters extends SubmitParams {
+    private final Boolean permanent;
+
+    public SessionParameters(Boolean permanent, String name, String file, String master, String mainClass, Integer numExecutors, Integer executorCores,
```
I think we need to add `@Nullable` annotations here, as we do on `SubmitParams`, otherwise Micronaut validation fails (I'm not completely sure about it)

pdambrauskas replied:
`@MapFormat(transformation = FLAT, keyFormat = RAW)` should also be added

jmilkiewicz replied:
done

jmilkiewicz commented on server/src/main/java/com/exacaster/lighter/rest/SessionController.java:
```diff
@@ -62,8 +62,15 @@ public Object get(@QueryValue(defaultValue = "0") Integer from,
+    @Put("/{id}")
+    @Status(HttpStatus.CREATED)
+    public Application create2(@PathVariable String id, @Valid @Body SessionParameters session) {
+        return sessionService.createSession(id, session);
```
I wanna to return 409 conflict given session with given ID exists. WDYT?

jmilkiewicz replied:
done

pdambrauskas commented on server/src/main/java/com/exacaster/lighter/application/sessions/SessionHandler.java:
```diff
@@ -114,6 +162,7 @@ public void handleTimeout() {
         if (timeoutInterval != null && !timeoutInterval.isZero()) {
             sessionService.fetchRunning()
                     .stream()
+                    //TODO here we can delete is as we would only care about PERMANENT_SESSION
```
yes, you can do that :)

jmilkiewicz replied:
done

pdambrauskas commented on server/src/main/java/com/exacaster/lighter/application/sessions/SessionService.java:
```diff
@@ -31,31 +34,47 @@ public class SessionService {
         this.statementHandler = statementHandler;
     }
 
+    //TODO do we wanna return just Session or PermSession as well?
```
we should return both, We want to show permanent sessions on the Lighter UI as we do with regular ones. This method is used by sessions API endpoint.

jmilkiewicz replied:
done.

pdambrauskas replied:
Remove `//TODO` comment as well :)

pdambrauskas commented on server/src/main/java/com/exacaster/lighter/application/sessions/SessionService.java:
```diff
@@ -31,31 +36,51 @@ public class SessionService {
-    public Application createSession(SubmitParams params, String sessionId) {
+    public Application createSession(String sessionId, SessionParams sessionParams) {
+        //TODO do we wanna do it for all session or perm only
+        if (applicationStorage.findApplication(sessionId).isPresent()) {
```
> The easiest option for this case it to simply save new session with given id with SQL INSERT (not with try to UPDATE and then INSERT how it is done now in saveApplication method) and then catch exception on duplicate key. WDYT ?

Makes sense. In that case we should remove `saveApplication` and introduce two separate methods - `insertApplication` & `updateApplication`.

jmilkiewicz replied:
Or we can leave saveApplication as it is - maybe even change it's name to saveOrUpdate (like in JPA) and introduce new method like insertApplication

jmilkiewicz replied:
I decided to add an extra method

jmilkiewicz commented:
I consider it is mostly done. 
@pdambrauskas @Minutis have a look

pdambrauskas commented on server/src/main/java/com/exacaster/lighter/application/Application.java:
```diff
@@ -65,6 +68,14 @@ public SubmitParams getSubmitParams() {
+        return deleted;
+    }
+
+    public boolean isNotDeleted() {
```
I think `@JsonIgnore` should be added on this one. We do not want to have both "isDeleted" & "isNotDeleted"  on  the json representation of application, when we return it on API endpoints. 

It can be added on `isDeleted()` as well, I think there are no cases when we return deleted applications on the http api layer.

jmilkiewicz replied:
right

jmilkiewicz replied:
done

pdambrauskas commented on server/src/main/java/com/exacaster/lighter/application/sessions/SessionHandler.java:
```diff
@@ -51,28 +56,71 @@ public Waitable launch(Application application, Consumer<Throwable> errorHandler
+
+    }
+
+    private static class PermanentSessionParam {
```
Minor: Maybe it would be nicer for readability to move class definition to the bottom of the file, after all method definitions?

jmilkiewicz replied:
done.

pdambrauskas commented on server/src/main/java/com/exacaster/lighter/application/sessions/SessionService.java:
```diff
@@ -31,40 +35,56 @@ public class SessionService {
+    //TODO do we wanna return just Session or PermSession as well?
     public List<Application> fetch(Integer from, Integer size) {
-        return applicationStorage.findApplications(ApplicationType.SESSION, from, size);
+        return applicationStorage.findApplications(EnumSet.of(ApplicationType.SESSION, ApplicationType.PERMANENT_SESSION), from, size);
```
`EnumSet.of(ApplicationType.SESSION, ApplicationType.PERMANENT_SESSION)` could be move to a private constant - in that case we would not initiate a new EnumSet for each `fetch` invocation.

That constant could also be reused on `deleteOne`

jmilkiewicz replied:
done

pdambrauskas commented on server/src/main/resources/application.yml:
```diff
@@ -4,7 +4,7 @@ lighter:
     frontend-path: file:${FRONTEND_PATH:../frontend/build}
     py-gateway-port: 25333
-    url: http://lighter.spark:8080
+    url: http://localhost:8080
```
Lets not change it on the main application file, to run the lighter locally, there is a `application-local.yaml`, to activate it you can set env var `MICRONAUT_ENVIRONMENTS=local` or set run property `-Dmicronaut.environments=local`

jmilkiewicz replied:
sorry my bad

jmilkiewicz replied:
done

pdambrauskas commented on server/src/test/groovy/com/exacaster/lighter/storage/jdbc/JdbcApplicationStorageTest.groovy:
```diff
@@ -62,4 +65,43 @@ class JdbcApplicationStorageTest extends Specification {
+        storage.deleteApplication(savedPermanentSession.id)
+
+        then: "fetching apps ignores soft deleted ones"
+//        storage.findApplications(EnumSet.of(ApplicationType.PERMANENT_SESSION, ApplicationType.SESSION), 0, 10).size() == 1
```
lets remove this, if it is not needed

jmilkiewicz replied:
ok

jmilkiewicz replied:
done

pdambrauskas commented:
I've left some minor comments, other than that - nice job!

jmilkiewicz commented:
> I've left some minor comments, other than that - nice job!

fixed comments
